### PR TITLE
Human descriptor

### PR DESCRIPTION
**Pull request by matemax**

### CODE REVIEW

File: examples/sdk_examples/human_descriptor_estimation.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import pprint
+
+from lunavl.sdk.faceengine.engine import VLFaceEngine
+from lunavl.sdk.image_utils.image import VLImage
+from resources import EXAMPLE_O
+
+
+def estimateDescriptor():
+    """
+    Estimate face descriptor.
```
Comment (matemax) [inline examples/sdk_examples/human_descriptor_estimation.py]: `face`

File: examples/sdk_examples/human_warper.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import pprint
+
+from lunavl.sdk.faceengine.engine import VLFaceEngine
+from lunavl.sdk.image_utils.image import VLImage
+from resources import EXAMPLE_O
+
+
+def createWarp():
+    """
+    Create warp from detection.
```
Comment (matemax) [inline examples/sdk_examples/human_warper.py]: `human detection`

File: lunavl/sdk/estimators/body_estimators/humanwarper.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+
+class HumanWarpedImage(VLImage):
+    """
+    Raw warped image.
+
+    Properties of a warped image:
+
+        - it's always in RGB color format
+        - it always contains just a single face
```
Comment (matemax) [inline lunavl/sdk/estimators/body_estimators/humanwarper.py]: `single face`

File: lunavl/sdk/estimators/body_estimators/humanwarper.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        - the face is always centered and rotated so that imaginary line between the eyes is horizontal.
```
Comment (matemax) [inline lunavl/sdk/estimators/body_estimators/humanwarper.py]: `face`

File: lunavl/sdk/estimators/body_estimators/humanwarper.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+            this checks are not guarantee that image is warp. This function is intended for debug
+        """
+        if self.format != self.format.R8G8B8:
+            raise ValueError("Bad image format for warped image, must be R8G8B8")
+
+    #  pylint: disable=W0221
+    @classmethod
+    def load(cls, *, filename: Optional[str] = None, url: Optional[str] = None) -> "HumanWarpedImage":  # type: ignore
+        """
+        Load imag from numpy array or file or url.
```
Comment (matemax) [inline lunavl/sdk/estimators/body_estimators/humanwarper.py]: `imag`

File: lunavl/sdk/estimators/body_estimators/humanwarper.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        return self
+
+
+class HumanWarp:
+    """
+    Structure for storing warp.
+
+    Attributes:
+        sourceDetection (HumanDetection): detection which generated warp
+        warpedImage (FaceWarpedImage):
```
Comment (matemax) [inline lunavl/sdk/estimators/body_estimators/humanwarper.py]: `FaceWarpedImage`

File: lunavl/sdk/estimators/body_estimators/humanwarper.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        """
+        self._coreWarper = coreWarper
+
+    @CoreExceptionWrap(LunaVLError.CreationWarpError)
+    def warp(self, humanDetection: HumanDetection) -> HumanWarp:
+        """
+        Create warp from detection.
+
+        Args:
+            humanDetection: face detection with landmarks5
```
Comment (matemax) [inline lunavl/sdk/estimators/body_estimators/humanwarper.py]: `face`

File: lunavl/sdk/descriptors/descriptors.py [hunk 8]
```diff
@@ -93,14 +94,16 @@ def reload(self, descriptor: bytes, garbageScore: float = 0.0) -> None:
         self.garbageScore = garbageScore
 
 
-class FaceDescriptorBatch(BaseEstimation):
+class BaseDescriptorBatch(BaseEstimation):
     """
     Face descriptor batch.
```
Comment (elenlee) [inline lunavl/sdk/descriptors/descriptors.py]:
```suggestion
    Base descriptor batch.
```

File: lunavl/sdk/descriptors/descriptors.py [hunk 9]
```diff
@@ -186,7 +189,7 @@ def __repr__(self) -> str:
         return str(fullDescriptors)
 
 
-class FaceDescriptorFactory:
+class BaseDescriptorFactory:
     """
     Face Descriptor factory.
```
Comment (elenlee) [inline lunavl/sdk/descriptors/descriptors.py]:
```suggestion
    Base Descriptor factory.
```

File: lunavl/sdk/estimators/body_estimators/human_descriptor.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+        """
+        super().__init__(coreExtractor)
+        self.descriptorFactory = humanDescriptorFactory
+
+    #  pylint: disable=W0221
+    def estimate(  # type: ignore
+        self, warp: Union[HumanWarp, HumanWarpedImage], descriptor: Optional[HumanDescriptor] = None
+    ) -> HumanDescriptor:
+        """
+        Estimate face descriptor from a warp image.
```
Comment (elenlee) [inline lunavl/sdk/estimators/body_estimators/human_descriptor.py]:
```suggestion
        Estimate human descriptor from a warp image.
```

File: lunavl/sdk/descriptors/descriptors.py [hunk 11]
```diff
@@ -4,17 +4,18 @@
 See `face descriptor`_.
```
Comment (elenlee) [inline lunavl/sdk/descriptors/descriptors.py:4]: sphinx документации нет

File: lunavl/sdk/descriptors/descriptors.py [hunk 12]
```diff
@@ -148,20 +151,20 @@ def __getitem__(self, i) -> FaceDescriptor:
...
+    def __iter__(self) -> Iterator[BaseDescriptor]:
         """
         Iterator by batch.
 
         Returns:
             iterator by descriptors.
         """
         for index in range(len(self)):
-            yield FaceDescriptor(self._coreEstimation.getDescriptorFast(index), self.scores[index])
+            yield self.__class__._descriptorFactory(self._coreEstimation.getDescriptorFast(index), self.scores[index])
```
Comment (Zeskbest) [inline lunavl/sdk/descriptors/descriptors.py]: What for? Why won't work without black magic?
```suggestion
            yield self._descriptorFactory(self._coreEstimation.getDescriptorFast(index), self.scores[index])
```
Search for other `__class__` also pls.

Comment (Zeskbest) [reply lunavl/sdk/descriptors/descriptors.py]: also no `__class__` usage [here](https://github.com/matemax/lunasdk/pull/82/files#diff-4892cfe567cd5f0b37e22f0c241672d7R269)

File: lunavl/sdk/estimators/body_estimators/human_descriptor.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        warps: List[Union[HumanWarp, HumanWarpedImage]],
+        aggregate: bool = False,
+        descriptorBatch: Optional[HumanDescriptorBatch] = None,
+    ) -> Tuple[HumanDescriptorBatch, Union[HumanDescriptor, None]]:
+        """
+        Estimate a batch of descriptors from warped images.
+
+        Args:
+            warps: warped images
+            aggregate:  whether to estimate  aggregate descriptor or not
```
Comment (Zeskbest) [inline lunavl/sdk/estimators/body_estimators/human_descriptor.py]:
```suggestion
            aggregate: whether to estimate an aggregated descriptor or not
```

File: lunavl/sdk/estimators/body_estimators/human_descriptor.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+            descriptorBatch: optional batch for saving descriptors
+
+        Returns:
+            tuple of batch and the aggregate descriptors (or None)
```
Comment (Zeskbest) [inline lunavl/sdk/estimators/body_estimators/human_descriptor.py]:
```suggestion
            tuple with a batch and a aggregated descriptor (or None)
```

File: lunavl/sdk/estimators/body_estimators/humanwarper.py [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        """
+        self._coreWarper = coreWarper
+
+    @CoreExceptionWrap(LunaVLError.CreationWarpError)
+    def warp(self, humanDetection: HumanDetection) -> HumanWarp:
+        """
+        Create warp from detection.
+
+        Args:
+            humanDetection: human body detection with landmarks5
```
Comment (Zeskbest) [inline lunavl/sdk/estimators/body_estimators/humanwarper.py]: ?
```suggestion
            humanDetection: human body detection with landmarks17
```

File: lunavl/sdk/estimators/estimators_utils/extractor_utils.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def estimate(
+    warp: Union[HumanWarp, HumanWarpedImage, FaceWarp, FaceWarpedImage],
+    descriptorFactory: BaseDescriptorFactory,
+    coreEstimator: IDescriptorExtractorPtr,
+    descriptor: Optional[BaseDescriptor] = None,
+) -> BaseDescriptor:
+    """
+    Estimate face descriptor from a warp image.
```
Comment (Zeskbest) [inline lunavl/sdk/estimators/estimators_utils/extractor_utils.py]:
```suggestion
    Estimate a face descriptor or a human descriptor from the warped image.
```

File: lunavl/sdk/faceengine/engine.py [hunk 17]
```diff
@@ -226,3 +229,29 @@ def createHumanDetector(self) -> HumanDetector:
             detector
         """
         return HumanDetector(self._faceEngine.createHumanDetector())
+
+    def createHumanWarper(self) -> HumanWarper:
+        return HumanWarper(self._faceEngine.createHumanWarper())
```
Comment (Zeskbest) [inline lunavl/sdk/faceengine/engine.py:241]: add a comment pls

File: tests/test_estimate_descriptor.py [hunk 18]
```diff
@@ -2,243 +2,410 @@
...
+from lunavl.sdk.estimators.body_estimators.humanwarper import HumanWarpedImage
+from lunavl.sdk.estimators.face_estimators.face_descriptor import FaceDescriptorEstimator
+from lunavl.sdk.estimators.face_estimators.facewarper import FaceWarpedImage
+from lunavl.sdk.globals import DEFAULT_HUMAN_DESCRIPTOR_VERSION as DHDV
 from tests.base import BaseTestClass
-from tests.resources import WARP_WHITE_MAN
+from tests.resources import WARP_WHITE_MAN, HUMAN_WARP
 
-EDVa = EXISTENT_DESCRIPTOR_VERSION_ABUNDANCE = [46, 52, 54, 56]
+# from unittest.case import _SubTest
```
Comment (Zeskbest) [inline tests/test_estimate_descriptor.py]:
```suggestion
```

File: tests/test_estimate_descriptor.py [hunk 19]
```diff
@@ -2,243 +2,410 @@
...
+        cls.humanDescriptor = cls.humanEstimator.estimate(humanWarp)
+        cls.humanDescriptorBatch, cls.aggregatedHumanDescriptor = estimator.estimateDescriptorsBatch(
+            humanWarps, aggregate=True
+        )
+
+    def descriptorSubTest(
+        self,
+    ) -> Generator[Tuple[ContextManager[None], DescriptorCase], Tuple[ContextManager[None], DescriptorCase], None]:
+        """
+        Generator for sub tests from root and user
```
Comment (Zeskbest) [inline tests/test_estimate_descriptor.py]:
```suggestion
        Generator for sub tests for human descriptor and face descriptor.
```

File: tests/test_estimate_descriptor.py [hunk 20]
```diff
@@ -2,243 +2,410 @@
...
+        for case in self.cases:
+            with self.subTest(type=case.type):
+                for planVersion in case.versions:
+
+                    with self.subTest(descriptor_version=planVersion):
+                        try:
+                            case.extractorFactory(descriptorVersion=planVersion)
+                        except RuntimeError as e:
+                            raise AssertionError(
+                                f"Descriptor version {planVersion} i" f"s not supported. But must be."
```
Comment (Zeskbest) [inline tests/test_estimate_descriptor.py]:
```suggestion
                                f"Descriptor version {planVersion} is not supported. But must be."
```

File: tests/test_estimate_descriptor.py [hunk 21]
```diff
@@ -2,243 +2,410 @@
...
+                            ) from e
 
     @pytest.mark.skip("need support 57 version")
```
Comment (Zeskbest) [inline tests/test_estimate_descriptor.py]: use self.skip instead?

File: lunavl/sdk/descriptors/descriptors.py (hunk 11 above)
Comment (matemax) [reply lunavl/sdk/descriptors/descriptors.py:4]: add later. docs out of date

File: lunavl/sdk/faceengine/engine.py (hunk 17 above)
Comment (matemax) [reply lunavl/sdk/faceengine/engine.py:241]: added

File: tests/test_estimate_descriptor.py (hunk 21 above)
Comment (matemax) [reply tests/test_estimate_descriptor.py]: ++ done

**Pull request merged**